zerver/migrations/0526_presencesequence.py

```diff
@@ -0,0 +1,73 @@
+# Generated by Django 5.0.5 on 2024-05-02 22:36
+
+import django.db.models.deletion
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def create_presence_sequence_for_old_realms(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    PresenceSequence = apps.get_model("zerver", "PresenceSequence")
+
+    max_id = Realm.objects.aggregate(models.Max("id"))["id__max"]
+    if max_id is None:
+        # Nothing to do if there are no rows yet.
+        return
+
+    BATCH_SIZE = 2000
+    lower_bound = 0
+
+    # Make sure we run past the end in case of new rows created while we run.
+    max_id += BATCH_SIZE / 2
+
+    while lower_bound < max_id:
+        realm_ids = Realm.objects.filter(
+            id__gt=lower_bound,
+            id__lte=lower_bound + BATCH_SIZE,
+            # Filter to realm whose PresenceSequence does not exist, to avoid
+            # running into IntegrityError by trying to create duplicate PresenceSequence.
+            presencesequence=None,
+        ).values_list("id", flat=True)
+
+        PresenceSequence.objects.bulk_create(
+            PresenceSequence(realm_id=realm_id, last_update_id=0) for realm_id in realm_ids
+        )
```

Comment:
Do we need to include an `on_conflict` specification, in case another process adds a presence sequence row while this is running?

Reply:
There shouldn't be any way for that to happen. The code ensures to create `PresenceSequence` whenever it's creating a new realm, but doesn't touch old realms. Old realms are assumed in the later commit to just correctly have it (via the backfill migrations).

In particular, the `do_update_user_presence` logic intentionally doesn't try to create `PresenceSequence` for a realm if it's somehow missing - it would complicate reasoning about the queries and locking situation there.

`create_realm` creates the `PresenceSequence` in `transaction_atomic()` together with the `Realm` row creation, so there should also be no way for this migration to grab a realm in a not-yet-complete state either.

Reply:
OK, that makes sense. But what this does mean is that presence API requests will 500 until this migration is complete... or maybe most correctly their queue may backlog? We can work this out in the deployment planning thread, which will include how to rework the commits in this PR.